### ISSUE TITLE
EPS test failures with 9.0.x

### ISSUE THREAD

**Issue by thesamesam** [python-pillow/Pillow#6013]
### What did you do?

Ran Pillow 9.0.1's test suite, although Pillow 9.0.0 fails in the same way.

### What did you expect to happen?

All tests pass.

### What actually happened?

Test failures in:
- Tests/test_file_eps.py::test_showpage
- Tests/test_file_eps.py::test_bytesio_object 
- Tests/test_file_eps.py::test_render_scale1
- Tests/test_file_eps.py::test_render_scale2

```
FAILED Tests/test_file_eps.py::test_showpage - AssertionError:  average pixel value difference 489.6932 > epsilon 6.0000
FAILED Tests/test_file_eps.py::test_bytesio_object - AssertionError:  average pixel value difference 498.7524 > epsilon 5.0000
FAILED Tests/test_file_eps.py::test_render_scale1 - AssertionError:  average pixel value difference 498.7524 > epsilon 5.0000
FAILED Tests/test_file_eps.py::test_render_scale2 - AssertionError:  average pixel value difference 503.5809 > epsilon 5.0000
==================================================================== 4 failed, 2642 passed, 231 skipped, 4 xfailed, 5 warnings in 41.82s =====================================================================
 * pytest failed with python3.8
 * ERROR: dev-python/pillow-9.0.1::gentoo failed (test phase):
 *   Failed to run 'epytest -p no:relaxed'
```

In more detail:
<pre>
================================================================================================== FAILURES ==================================================================================================
_______________________________________________________________________________________________ test_showpage ________________________________________________________________________________________________

    @pytest.mark.skipif(not HAS_GHOSTSCRIPT, reason="Ghostscript not available")
    def test_showpage():
        # See https://github.com/python-pillow/Pillow/issues/2615
        with Image.open("Tests/images/reqd_showpage.eps") as plot_image:
            with Image.open("Tests/images/reqd_showpage.png") as target:
                # should not crash/hang
                plot_image.load()
                #  fonts could be slightly different
>               assert_image_similar(plot_image, target, 6)

plot_image = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=553x475 at 0x7F24E34C9C40>
target     = <PIL.PngImagePlugin.PngImageFile image mode=RGB size=553x475 at 0x7F24E1DA9310>

Tests/test_file_eps.py:96:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
Tests/helper.py:134: in assert_image_similar
    raise e
        a          = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=553x475 at 0x7F24E34C9C40>
        ach        = <PIL.Image.Image image mode=L size=553x475 at 0x7F24E1DA94F0>
        ave_diff   = 489.6931950128486
        b          = <PIL.PngImagePlugin.PngImageFile image mode=RGB size=553x475 at 0x7F24E1DA9310>
        bch        = <PIL.Image.Image image mode=L size=553x475 at 0x7F24E1DA9520>
        chdiff     = <PIL.Image.Image image mode=L size=553x475 at 0x7F24E2FF24C0>
        diff       = 128630160
        epsilon    = 6
        msg        = None
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=553x475 at 0x7F24E34C9C40>, b = <PIL.PngImagePlugin.PngImageFile image mode=RGB size=553x475 at 0x7F24E1DA9310>, epsilon = 6, msg = None

    def assert_image_similar(a, b, epsilon, msg=None):
        assert a.mode == b.mode, msg or f"got mode {repr(a.mode)}, expected {repr(b.mode)}"
        assert a.size == b.size, msg or f"got size {repr(a.size)}, expected {repr(b.size)}"

        a, b = convert_to_comparable(a, b)

        diff = 0
        for ach, bch in zip(a.split(), b.split()):
            chdiff = ImageMath.eval("abs(a - b)", a=ach, b=bch).convert("L")
            diff += sum(i * num for i, num in enumerate(chdiff.histogram()))

        ave_diff = diff / (a.size[0] * a.size[1])
        try:
>           assert epsilon >= ave_diff, (
                (msg or "")
                + f" average pixel value difference {ave_diff:.4f} > epsilon {epsilon:.4f}"
            )
E           AssertionError:  average pixel value difference 489.6932 > epsilon 6.0000
E           assert 6 >= 489.6931950128486

a          = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=553x475 at 0x7F24E34C9C40>
ach        = <PIL.Image.Image image mode=L size=553x475 at 0x7F24E1DA94F0>
ave_diff   = 489.6931950128486
b          = <PIL.PngImagePlugin.PngImageFile image mode=RGB size=553x475 at 0x7F24E1DA9310>
bch        = <PIL.Image.Image image mode=L size=553x475 at 0x7F24E1DA9520>
chdiff     = <PIL.Image.Image image mode=L size=553x475 at 0x7F24E2FF24C0>
diff       = 128630160
epsilon    = 6
msg        = None

Tests/helper.py:123: AssertionError
____________________________________________________________________________________________ test_bytesio_object _____________________________________________________________________________________________

    @pytest.mark.skipif(not HAS_GHOSTSCRIPT, reason="Ghostscript not available")
    def test_bytesio_object():
        with open(FILE1, "rb") as f:
            img_bytes = io.BytesIO(f.read())

        with Image.open(img_bytes) as img:
            img.load()

            with Image.open(FILE1_COMPARE) as image1_scale1_compare:
                image1_scale1_compare = image1_scale1_compare.convert("RGB")
            image1_scale1_compare.load()
>           assert_image_similar(img, image1_scale1_compare, 5)

f          = <_io.BufferedReader name='Tests/images/zero_bb.eps'>
image1_scale1_compare = <PIL.Image.Image image mode=RGB size=460x352 at 0x7F24E3432BE0>
img        = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=460x352 at 0x7F24E3432160>
img_bytes  = <_io.BytesIO object at 0x7f24e20fc9a0>

Tests/test_file_eps.py:137:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
Tests/helper.py:134: in assert_image_similar
    raise e
        a          = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=460x352 at 0x7F24E3432160>
        ach        = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E34328E0>
        ave_diff   = 498.75240859683794
        b          = <PIL.Image.Image image mode=RGB size=460x352 at 0x7F24E3432BE0>
        bch        = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E34DDBE0>
        chdiff     = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E3223D60>
        diff       = 80757990
        epsilon    = 5
        msg        = None
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=460x352 at 0x7F24E3432160>, b = <PIL.Image.Image image mode=RGB size=460x352 at 0x7F24E3432BE0>, epsilon = 5, msg = None

    def assert_image_similar(a, b, epsilon, msg=None):
        assert a.mode == b.mode, msg or f"got mode {repr(a.mode)}, expected {repr(b.mode)}"
        assert a.size == b.size, msg or f"got size {repr(a.size)}, expected {repr(b.size)}"

        a, b = convert_to_comparable(a, b)

        diff = 0
        for ach, bch in zip(a.split(), b.split()):
            chdiff = ImageMath.eval("abs(a - b)", a=ach, b=bch).convert("L")
            diff += sum(i * num for i, num in enumerate(chdiff.histogram()))

        ave_diff = diff / (a.size[0] * a.size[1])
        try:
>           assert epsilon >= ave_diff, (
                (msg or "")
                + f" average pixel value difference {ave_diff:.4f} > epsilon {epsilon:.4f}"
            )
E           AssertionError:  average pixel value difference 498.7524 > epsilon 5.0000
E           assert 5 >= 498.75240859683794

a          = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=460x352 at 0x7F24E3432160>
ach        = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E34328E0>
ave_diff   = 498.75240859683794
b          = <PIL.Image.Image image mode=RGB size=460x352 at 0x7F24E3432BE0>
bch        = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E34DDBE0>
chdiff     = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E3223D60>
diff       = 80757990
epsilon    = 5
msg        = None

Tests/helper.py:123: AssertionError
_____________________________________________________________________________________________ test_render_scale1 _____________________________________________________________________________________________

    @pytest.mark.skipif(not HAS_GHOSTSCRIPT, reason="Ghostscript not available")
    @skip_unless_feature("zlib")
    def test_render_scale1():
        # We need png support for these render test

        # Zero bounding box
        with Image.open(FILE1) as image1_scale1:
            image1_scale1.load()
            with Image.open(FILE1_COMPARE) as image1_scale1_compare:
                image1_scale1_compare = image1_scale1_compare.convert("RGB")
            image1_scale1_compare.load()
>           assert_image_similar(image1_scale1, image1_scale1_compare, 5)

image1_scale1 = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=460x352 at 0x7F24E349ABE0>
image1_scale1_compare = <PIL.Image.Image image mode=RGB size=460x352 at 0x7F24E349A820>

Tests/test_file_eps.py:158:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
Tests/helper.py:134: in assert_image_similar
    raise e
        a          = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=460x352 at 0x7F24E349ABE0>
        ach        = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E349AA90>
        ave_diff   = 498.75240859683794
        b          = <PIL.Image.Image image mode=RGB size=460x352 at 0x7F24E349A820>
        bch        = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E349A5E0>
        chdiff     = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E349A970>
        diff       = 80757990
        epsilon    = 5
        msg        = None
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=460x352 at 0x7F24E349ABE0>, b = <PIL.Image.Image image mode=RGB size=460x352 at 0x7F24E349A820>, epsilon = 5, msg = None

    def assert_image_similar(a, b, epsilon, msg=None):
        assert a.mode == b.mode, msg or f"got mode {repr(a.mode)}, expected {repr(b.mode)}"
        assert a.size == b.size, msg or f"got size {repr(a.size)}, expected {repr(b.size)}"

        a, b = convert_to_comparable(a, b)

        diff = 0
        for ach, bch in zip(a.split(), b.split()):
            chdiff = ImageMath.eval("abs(a - b)", a=ach, b=bch).convert("L")
            diff += sum(i * num for i, num in enumerate(chdiff.histogram()))

        ave_diff = diff / (a.size[0] * a.size[1])
        try:
>           assert epsilon >= ave_diff, (
                (msg or "")
                + f" average pixel value difference {ave_diff:.4f} > epsilon {epsilon:.4f}"
            )
E           AssertionError:  average pixel value difference 498.7524 > epsilon 5.0000
E           assert 5 >= 498.75240859683794

a          = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=460x352 at 0x7F24E349ABE0>
ach        = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E349AA90>
ave_diff   = 498.75240859683794
b          = <PIL.Image.Image image mode=RGB size=460x352 at 0x7F24E349A820>
bch        = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E349A5E0>
chdiff     = <PIL.Image.Image image mode=L size=460x352 at 0x7F24E349A970>
diff       = 80757990
epsilon    = 5
msg        = None

Tests/helper.py:123: AssertionError
_____________________________________________________________________________________________ test_render_scale2 _____________________________________________________________________________________________

    @pytest.mark.skipif(not HAS_GHOSTSCRIPT, reason="Ghostscript not available")
    @skip_unless_feature("zlib")
    def test_render_scale2():
        # We need png support for these render test

        # Zero bounding box
        with Image.open(FILE1) as image1_scale2:
            image1_scale2.load(scale=2)
            with Image.open(FILE1_COMPARE_SCALE2) as image1_scale2_compare:
                image1_scale2_compare = image1_scale2_compare.convert("RGB")
            image1_scale2_compare.load()
>           assert_image_similar(image1_scale2, image1_scale2_compare, 5)

image1_scale2 = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=920x704 at 0x7F24E1DA96A0>
image1_scale2_compare = <PIL.Image.Image image mode=RGB size=920x704 at 0x7F24E1DA91C0>

Tests/test_file_eps.py:180:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
Tests/helper.py:134: in assert_image_similar
    raise e
        a          = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=920x704 at 0x7F24E1DA96A0>
        ach        = <PIL.Image.Image image mode=L size=920x704 at 0x7F24E1DA96D0>
        ave_diff   = 503.5809041501976
        b          = <PIL.Image.Image image mode=RGB size=920x704 at 0x7F24E1DA91C0>
        bch        = <PIL.Image.Image image mode=L size=920x704 at 0x7F24E1DA9160>
        chdiff     = <PIL.Image.Image image mode=L size=920x704 at 0x7F24E34C9400>
        diff       = 326159280
        epsilon    = 5
        msg        = None
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

a = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=920x704 at 0x7F24E1DA96A0>, b = <PIL.Image.Image image mode=RGB size=920x704 at 0x7F24E1DA91C0>, epsilon = 5, msg = None

    def assert_image_similar(a, b, epsilon, msg=None):
        assert a.mode == b.mode, msg or f"got mode {repr(a.mode)}, expected {repr(b.mode)}"
        assert a.size == b.size, msg or f"got size {repr(a.size)}, expected {repr(b.size)}"

        a, b = convert_to_comparable(a, b)

        diff = 0
        for ach, bch in zip(a.split(), b.split()):
            chdiff = ImageMath.eval("abs(a - b)", a=ach, b=bch).convert("L")
            diff += sum(i * num for i, num in enumerate(chdiff.histogram()))

        ave_diff = diff / (a.size[0] * a.size[1])
        try:
>           assert epsilon >= ave_diff, (
                (msg or "")
                + f" average pixel value difference {ave_diff:.4f} > epsilon {epsilon:.4f}"
            )
E           AssertionError:  average pixel value difference 503.5809 > epsilon 5.0000
E           assert 5 >= 503.5809041501976

a          = <PIL.EpsImagePlugin.EpsImageFile image mode=RGB size=920x704 at 0x7F24E1DA96A0>
ach        = <PIL.Image.Image image mode=L size=920x704 at 0x7F24E1DA96D0>
ave_diff   = 503.5809041501976
b          = <PIL.Image.Image image mode=RGB size=920x704 at 0x7F24E1DA91C0>
bch        = <PIL.Image.Image image mode=L size=920x704 at 0x7F24E1DA9160>
chdiff     = <PIL.Image.Image image mode=L size=920x704 at 0x7F24E34C9400>
diff       = 326159280
epsilon    = 5
msg        = None

Tests/helper.py:123: AssertionError
</pre>

Full log available here: [build.log](https://github.com/python-pillow/Pillow/files/7991812/build.log)

### What are your OS, Python and Pillow versions?

* OS: Gentoo Linux
* Python: Python 3.8.12
* Pillow: 9.0.1

**Comment by radarhere** [comment]
It's weird - this happened in Arch Linux in https://github.com/python-pillow/docker-images/pull/134#issuecomment-1026574790 2 days ago, but fixed itself after rerunning just as mysteriously a day later.

**Comment by thesamesam** [comment]
> It's weird - this happened in Arch Linux in [python-pillow/docker-images#134 (comment)](https://github.com/python-pillow/docker-images/pull/134#issuecomment-1026574790) 2 days ago, but fixed itself after rerunning just as mysteriously a day later.

Weird! Thanks for the background, I feel a bit better knowing it's not Gentoo-specific now.

My Ghostscript version is `GPL Ghostscript 9.55.0 (2021-09-27)`. I did see about trying some of Debian's stack of patches (the 02* ones at https://sources.debian.org/src/ghostscript/9.55.0~dfsg-3/debian/patches/) but it didn't make any difference.

**Comment by radarhere** [comment]
If I try and replicate this [in Gentoo using Docker](https://github.com/radarhere/docker-images/tree/gentoo) with Ghostscript 9.55.0 from app-text/ghostscript-gpl, I don't see [any failing tests](https://github.com/radarhere/docker-images/runs/5088401383#step:5:32818).

Do you any ideas on how I can replicate this?

**Comment by thesamesam** [comment]
Hmm. Here's how we [build](https://gitweb.gentoo.org/repo/gentoo.git/tree/dev-python/pillow/pillow-9.0.1.ebuild) Pillow.

Let's start with seeing if, when using the ebuild, your Docker setup hits my failure or not? Could you try running (instead of `pip` for a moment):
`emerge -v1 --with-test-deps dev-python/pillow ; FEATURES=test emerge -v1 dev-python/pillow`?

**Comment by radarhere** [comment]
Ok, this took a while to run, but still, [no failing tests.](https://github.com/radarhere/docker-images/runs/5107231148)

```
FROM gentoo/stage3

RUN echo 'USE="xvfb X elogind"' | cat >> /etc/portage/make.conf
RUN echo 'FEATURES="-ipc-sandbox -pid-sandbox -network-sandbox -usersandbox -mount-sandbox -sandbox"' | cat >> /etc/portage/make.conf
RUN cat /etc/portage/make.conf
RUN emerge --quiet --sync || true
RUN emerge -v1 dev-lang/rust-bin
RUN emerge media-libs/libjpeg-turbo sys-libs/zlib app-text/ghostscript-gpl
RUN USE="jpeg jpeg2k tiff truetype xvfb X elogind" emerge -v1 --with-test-deps dev-python/pillow
RUN FEATURES=test USE="jpeg jpeg2k tiff truetype xvfb X elogind" emerge -v1 dev-python/pillow
```

**Comment by radarhere** [comment]
This has started occurring on my own machine now. It seems that the Ghostscript background color has changed.

I've put together https://github.com/radarhere/Pillow/commit/8a7ca7a06e0028407f314880ce48725557f6af35 to fix my problem. Does that fix your problem as well?

**Comment by thesamesam** [comment]
@radarhere Happy to try it. Will do so shortly. Any idea what caused the change? Did you just get the new GS version or something else?

**Comment by thesamesam** [comment]
This feels like a cruel joke at this point: I can't reproduce pre-patch now. It passes for me now before and after patch.

I've asked some of the Gentoo Python team to try it out and let me know if they can hit the failure. It's definitely something odd about the environment because when I reported this bug, 9.0.0 failed too, even though it didn't when it was released!

**Comment by radarhere** [comment]
Yep, same here! If I upgrade to lcms 2.13.1, the problem is fixed. Thanks for figuring that out.

**Comment by radarhere** [comment]
lcms2 2.13.1 was released 5 days after 2.13.
That's a pretty good response time on their part, so I imagine this would affect a relatively small number of people, and I would hope that it's not heavily cached by package managers.
Those who do notice it, like Pillow's docker-images, yourself and myself, might easily find the problem fixed by an upgrade to lcms2 2.13.1 well before Pillow 9.1.0 is released on April 1.

Are you happy to consider this resolved?

**Comment by thesamesam** [comment]
Completely happy to consider it resolved. Agreed, and if on the off chance someone pops up with this failure, we know what to tell them.
